query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala

```diff
@@ -395,75 +419,214 @@ final case class LabelCardinalityExec(queryContext: QueryContext,
+    require(extendToSize >= initPrefix.size,
+      "extendToSize must be at least initPrefix.size")
+
+    // A queue for each set of "space" (i.e. namespace, workspace, metric) labels.
+    // Spaces are stored in descending precedence, and each successive queue
+    //   contains labels for the "subspace" of the label at the front of the
+    //   previous queue.
+    private val spaceQueues = new Array[mutable.Queue[String]](extendToSize)
+
+    // Stores the prefix returned by next().
+    private val currPrefix = new mutable.ArrayBuffer[String](extendToSize)
+
+    private var iFirstEmptyQueue = -1
+
+    // TODO(a_theimer): this feels gross
+    initialize()
+
+    private def initialize(): Unit = {
+      // initialize each of the queues
+      for (i <- 0 until extendToSize) {
+        spaceQueues(i) = new mutable.Queue[String]
+      }
+
+      for (i <- 0 until initPrefix.size) {
+        // Place each existing prefix label into appropriate queue.
+        spaceQueues(i).enqueue(initPrefix(i))
+        // Store as much of the prefix as we have available.
+        currPrefix.append(initPrefix(i))
+      }
+
+      iFirstEmptyQueue = initPrefix.size
+
+      // Need this because an empty prefix is a valid argument, and hasNext()
+      //   relies on spaceQueues(0).size.
+      // This results in one extra enqueueAndUpdatePrefix() call at the first next().
+      enqueueAndUpdatePrefix()
+    }
+
+    /**
+     * Dequeues from each queue-- in ascending precedence-- until a dequeue()
+     *   does not leave a queue empty.
+     *
+     * This method mostly exists as setup for enqueueAndUpdatePrefix().
+     */
+    private def dequeue(): Unit = {
+      require(spaceQueues.last.nonEmpty, "should never dequeue without full prefix")
+      iFirstEmptyQueue = spaceQueues.size
+      for (queue <- spaceQueues.reverseIterator) {
+        queue.dequeue()
+        if (queue.nonEmpty) {
+          return
+        }
+        iFirstEmptyQueue = iFirstEmptyQueue - 1
+      }
+    }
+
+    /**
+     * Given the prefix defined by the front of each queue,
+     *   repopulates empty queues with "subspace" labels.
+     */
+    private def enqueueAndUpdatePrefix(): Unit = {
+      if (iFirstEmptyQueue == extendToSize) {
+        // final queue still contains elements; no need to call topk
+        currPrefix(currPrefix.size - 1) = spaceQueues(currPrefix.size - 1).front
+      } else {
+        currPrefix.reduceToSize(iFirstEmptyQueue)
+        if (iFirstEmptyQueue > 0) {
+          // this space name is still unused
+          currPrefix(iFirstEmptyQueue - 1) = spaceQueues(iFirstEmptyQueue - 1).front
+        }
+        // Fill each successive queue with the topKCardinality result such that
+        //   its argument prefix is defined by the preceding queue front labels.
+        for (iempty <- iFirstEmptyQueue until extendToSize) {
+          // TODO(a_theimer): MAX_RESPONSE_SIZE works, but not exactly the same meaning
```

Comment:
TODO: probably just remove the TODO, since MAX_RESPONSE_SIZE is sufficient enough.